what-is-pair.org

```diff
@@ -0,0 +1,253 @@
+setup.
+
+We use tmate to provide the pair programming environment, but it is not
+required. In fact, it is not the first choice for many of us already using Pair,
+as it has a noticeable lag and freezes. Instead, we use the Pair tool
+~pair-ssh-instance~...which, in a roundabout way, ssh's into the Environment
+container and runs a custom attach call to attach to that tmate socket with tmux
+instead (haha, i think, this can definitely be corrected). ~pair-ssh-instance~
+is faster and more reliable than tmate though, and we could have it setup to
+start a tmux session instead that we attach to.
+
+In other words, there is an large amount of freedom given to people using
+environment. We offer heavily recommended suggestions, but few constraints.
+
+Also, there are flows that work great with Pair, but are not emphasised in its
+architecture. For example, running [[https://tilt.dev/][Tilt]] as you develop in cluster, so all
+changes are immediately deployed to the cluster and accessible via url. This is
+handy, and supported, but not advertised.
+
+** Our Emacs configuration
+Emacs is a major part of our Environment setup.  It can be broken into four layers:
+- Emacs the program
+- Doom Emacs, a framework for using and configuring Emacs
+- Our custom ii elisp packages, that extend emacs support for tmate, iterm, and
+  developing with postgres
+- Our custom doom configuration, which installs the above packages and a number
+  of others, in an idiosyncratic ii custom way.
+
+We also cache our current version of doom, as it can have a long install time and we
+want clusters to come up quickly.
+
+I bring up the layers as each need to be synced to make sure emacs works
+correctly. Emacs, doom emacs, our emacs packages, and our doom emacs config are
+each versioned, and if there is an update to one of them, without a coordinate
+update to the rest of them, then it is likely emacs will not work properly.
+Because of this, we've been conservative with updates to our custom
+configuration.
+
+The biggest difference in our configuration is the support of tmate blocks in org-mode,
+and to have these blocks open up appropriate windows in the Environment tmate session. This
+was developed so we could write executable documents in cluster, and see the output in our
+shared tmate session.
+
+* Comparisons to other solutions
+** Kind and Minikube
+Kind and Minikube are two ways to quickly run kubernetes clusters, and are often
+recommended for trying out kubernetes or trying out some new tool for kubernetes.
+
+You can compare them to Pair as all offer a simple way to create a cluster relatively quickly.
+
+The biggest differences are:
+- Pair is a production level cluster, running on bare metal equinix servers.  Kind and minikube
+  run inside docker and a virtual box respectively.  They cannot simulate a production-grade environment.
+- Pair can be easily setup as a multi-node cluster.  This is impossible with Minikube and difficult with Kind.
```

Comment:
I think kind does multi-node fairly straightforward, @heyste were you having trouble with kind and multiple control-plane nodes?

Reply:
Yes, I did but not now. It ready easy to add multiple control-;plane nodes to the cluster config but what I didn't notice was the automatic creation of an `external-load-balancer` node. The ip of the load blancer will be set in the `kubeconfig`. The only down side with kind is you can't add extra nodes to a running cluster from the comments I've seen in the slack channel.

Reply:
clarified in https://github.com/sharingio/org/pull/1/commits/a37c1b5b38cfc4412c3b1f36ecaf7ef2e743c81c